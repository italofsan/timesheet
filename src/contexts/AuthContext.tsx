import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import { auth, firebase } from "../services/firebase";
import { errorMessage } from "../components/Messages";

type User = {
  id: string;
  name: string;
};

type AuthContextType = {
  user: User;
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
  isSigned: boolean;
  setIsSigned: Dispatch<SetStateAction<boolean>>;
};

type AuthContextProviderProps = {
  children: ReactNode;
};

export const AuthContext = createContext({} as AuthContextType);

export function AuthContextProvider(props: AuthContextProviderProps) {
  const [user, setUser] = useState<User>({
    id: "",
    name: "",
  });
  const [isSigned, setIsSigned] = useState<boolean>(false);

  function loadStorageData() {
    const storageIsSigned = localStorage.getItem("@SIGNED");

    if (storageIsSigned) {
      setIsSigned(JSON.parse(storageIsSigned));
    }
  }

  useEffect(() => {
    loadStorageData();
    const unsubscribe = auth.onAuthStateChanged((user) => {
      if (user) {
        const { displayName, uid } = user;

        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }

        setUser({
          id: uid,
          name: displayName,
        });
        setIsSigned(true);
        localStorage.setItem("@SIGNED", JSON.stringify(true));
      }
    });

    return () => {
      unsubscribe();
    };
  }, []);

  async function signInWithGoogle() {
    const provider = new firebase.auth.GoogleAuthProvider();
    try {
      const result = await auth.signInWithPopup(provider);
      if (result.user) {
        const { displayName, uid } = result.user;
        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }
        setUser({
          id: uid,
          name: displayName,
        });
      }
    } catch (error) {
      console.log(error);
      errorMessage("Não foi possível realizar Login!");
    }
  }

  async function signOut() {
    await auth.signOut();
    setUser({} as User);
    localStorage.clear();
    setIsSigned(false);
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        signInWithGoogle,
        signOut,
        isSigned,
        setIsSigned,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}
