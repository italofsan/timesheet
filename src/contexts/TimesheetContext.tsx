import {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from "react";
import { errorMessage, successMessage } from "../components/Messages";
import { DataTimesheetInput, DataTimesheet } from "../utils/types";
import {
  calcUndertimeHours,
  calcOvertimeHours,
  calcWorkedHours,
  sumHours,
} from "../utils/functions/mathHoursFunctions";
import { isTimesheetExists } from "../utils/functions/timesheetFunctions";
import { database } from "../services/firebase";
import { useAuth } from "../hooks/useAuth";

interface TimesheetContextProps {
  listTimesheets: DataTimesheet[];
  totalWorkedHours: string;
  totalOvertimeHours: string;
  totalUndertimeHours: string;
  handleAddTimesheet: (valuesForm: DataTimesheetInput) => void;
  handleUpdateTimesheet: (valuesForm: DataTimesheet) => void;
  handleDeleteTimesheet: (selectedTimesheetId: string) => void;
}

interface TimesheetContextProviderProps {
  children: ReactNode;
}

type FirebaseTimesheets = Record<string, DataTimesheet>;

export const TimesheetContext = createContext<TimesheetContextProps>(
  {} as TimesheetContextProps
);

export function TimesheetContextProvider({
  children,
}: TimesheetContextProviderProps) {
  const { user } = useAuth();
  const [listTimesheets, setListTimesheets] = useState<DataTimesheet[]>([]);
  const [totalWorkedHours, setTotalWorkedHours] = useState("00:00");
  const [totalOvertimeHours, setTotalOvertimeHours] = useState("00:00");
  const [totalUndertimeHours, setTotalUndertimeHours] = useState("00:00");

  useEffect(() => {
    setTotalWorkedHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.workedHours);
      }, "00:00")
    );

    setTotalOvertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.overtimeHours);
      }, "00:00")
    );

    setTotalUndertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.undertimeHours);
      }, "00:00")
    );
  }, [listTimesheets]);

  useEffect(() => {
    const timesheetRef = database.ref();
    timesheetRef.on("value", (timesheet) => {
      const databaseTimesheet = timesheet.val();
      const firebaseTimesheets: FirebaseTimesheets = databaseTimesheet
        ? databaseTimesheet.timesheets
        : {};

      if (firebaseTimesheets) {
        const parsedTimesheets = Object.entries(firebaseTimesheets)
          .map(([key, value]) => {
            return {
              id: key,
              date: value.date,
              idUser: value.idUser,
              arrivalTime: value.arrivalTime,
              lunchExitTime: value.lunchExitTime,
              lunchArrivalTime: value.lunchArrivalTime,
              exitTime: value.exitTime,
              undertimeHours: value.undertimeHours,
              overtimeHours: value.overtimeHours,
              workedHours: value.workedHours,
              createdAt: value.createdAt,
            };
          })
          .filter((timesheet) => timesheet.idUser === user.id);

        setListTimesheets(parsedTimesheets);
      }
    });

    return () => {
      timesheetRef.off("value");
    };
  }, [user.id]);

  const handleAddTimesheet = useCallback(
    async (valuesForm: DataTimesheetInput) => {
      if (isTimesheetExists(listTimesheets, valuesForm)) {
        return errorMessage("You already have a register for this day!");
      }
      const timesheetRef = database.ref("timesheets");
      try {
        await timesheetRef.push({
          ...valuesForm,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          createdAt: String(new Date()),
          date: String(new Date(valuesForm.date)),
          idUser: user.id,
        });
        successMessage("Register created gracefully!");
      } catch (error) {
        errorMessage("Unable to create register!");
      }
    },
    [user.id, listTimesheets]
  );

  const handleUpdateTimesheet = useCallback(
    async (valuesForm: DataTimesheet) => {
      try {
        await database.ref(`timesheets/${valuesForm.id}`).update({
          arrivalTime: valuesForm.arrivalTime,
          lunchExitTime: valuesForm.lunchExitTime,
          lunchArrivalTime: valuesForm.lunchArrivalTime,
          exitTime: valuesForm.exitTime,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
        });
        successMessage("Register updated gracefully!");
      } catch (error) {
        errorMessage("Unable to update register!");
      }
    },
    []
  );

  const handleDeleteTimesheet = useCallback(
    async (selectedTimesheetId: string) => {
      try {
        await database.ref(`timesheets/${selectedTimesheetId}`).remove();
        successMessage("Register removed gracefully!");
      } catch (error) {
        errorMessage("Unable to remove register!");
      }
    },
    []
  );

  return (
    <TimesheetContext.Provider
      value={{
        listTimesheets,
        totalWorkedHours,
        totalOvertimeHours,
        totalUndertimeHours,
        handleAddTimesheet,
        handleUpdateTimesheet,
        handleDeleteTimesheet,
      }}
    >
      {children}
    </TimesheetContext.Provider>
  );
}
