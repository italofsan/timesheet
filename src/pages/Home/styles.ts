import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: "flex",
      height: "100vh",
      background: "linear-gradient(0.25turn, rgb(143,125,238) , #FFF )",
    },
    appBar: {
      backgroundColor: colors.main,
    },
    title: {
      flexGrow: 1,
    },
    content: {
      flexGrow: 1,
      padding: theme.spacing(3),
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
    },
    imageContainer: {
      display: "flex",
      width: "100%",
      height: "100%",
      padding: "24px 0",
    },
    image: {
      width: "100%",
      maxHeight: 620,
    },
    loginWrapper: {
      display: "flex",
      justifyContent: "center",
    },
    loginContainer: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      padding: 48,
      maxWidth: 400,
    },
    loginTitle: {
      textAlign: "center",
      fontWeight: "bold",
      fontSize: 32,
      color: colors.main,
      paddingBottom: theme.spacing(3),
    },
    loginSubtitle: {
      paddingBottom: theme.spacing(3),
    },
    loginButton: {
      fontSize: 14,
      fontWeight: "bold",
      padding: "4px 24px",
      color: colors.white,
      transition: ".5s ease",
      backgroundColor: colors.main,
      "&:hover": {
        backgroundColor: colors.mainHover,
      },
      borderRadius: 5,
      textDecoration: "none",
    },
  })
);
