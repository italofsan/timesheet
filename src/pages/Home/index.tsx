import {
  AppBar,
  Container,
  Grid,
  Toolbar,
  Typography,
  Paper,
  Button,
} from "@material-ui/core";
import { useHistory } from "react-router-dom";

import timeHomePageImg from "../../assets/images/time-homepage.svg";

import { useAuth } from "../../hooks/useAuth";

import { useStyles } from "./styles";

export function Home() {
  const classes = useStyles();
  const history = useHistory();
  const { user, signInWithGoogle } = useAuth();

  async function handleLogin() {
    if (!user?.id) {
      await signInWithGoogle();
    }
    history.push("/app");
  }

  return (
    <div className={classes.root}>
      <AppBar position='fixed' className={classes.appBar}>
        <Container maxWidth='lg'>
          <Toolbar>
            <Typography variant='h6' className={classes.title}>
              Timesheet App
            </Typography>
          </Toolbar>
        </Container>
      </AppBar>
      <main className={classes.content}>
        <Container maxWidth='lg'>
          <Grid container justify='center' alignItems='center' direction='row'>
            <Grid item xs={12} sm={6} md={6} lg={6}>
              <div className={classes.imageContainer}>
                <img
                  src={timeHomePageImg}
                  alt='Time Home Page'
                  className={classes.image}
                />
              </div>
            </Grid>
            <Grid
              item
              xs={12}
              sm={6}
              md={6}
              lg={6}
              className={classes.loginWrapper}
            >
              <Paper className={classes.loginContainer}>
                <Typography className={classes.loginTitle}>Log In</Typography>
                <Typography className={classes.loginSubtitle}>
                  To log in to the application you need to use a Google account
                </Typography>
                <Button
                  size='small'
                  variant='contained'
                  className={classes.loginButton}
                  onClick={handleLogin}
                  role='loginButton'
                >
                  Login With Google
                </Button>
              </Paper>
            </Grid>
          </Grid>
        </Container>
      </main>
    </div>
  );
}
