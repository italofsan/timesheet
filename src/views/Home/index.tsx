import { useState } from "react";
import { Button, Container, Grid } from "@material-ui/core";

import { TimesheetTable } from "../../components/TimesheetTable";
import { AddModal } from "../../components/Modals/AddModal";
import { Summary } from "../../components/Summary";

import { useStyles } from "./styles";

export function Home() {
  const classes = useStyles();

  const [addModalOpen, setAddModalOpen] = useState(false);
  function handleOpenAddModal() {
    setAddModalOpen(true);
  }
  function handleCloseAddModal() {
    setAddModalOpen(false);
  }

  return (
    <Container maxWidth='lg' style={{ marginTop: 24 }}>
      <Grid container className={classes.root} spacing={2}>
        <Grid container spacing={4}>
          <Summary />
          <Grid item xs={12} className={classes.buttonContainer}>
            <Button
              size='small'
              variant='contained'
              className={classes.addButton}
              onClick={handleOpenAddModal}
            >
              + Add
            </Button>
          </Grid>
          <TimesheetTable />
        </Grid>

        <AddModal
          addModalOpen={addModalOpen}
          handleCloseAddModal={handleCloseAddModal}
        />
      </Grid>
    </Container>
  );
}
