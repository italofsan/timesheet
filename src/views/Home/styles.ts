import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      height: "100%",
    },
    addButton: {
      fontSize: 14,
      fontWeight: "bold",
      padding: "5px 25px",
      color: colors.white,
      transition: ".5s ease",
      backgroundColor: colors.main,
      "&:hover": {
        backgroundColor: colors.mainHover,
      },
      borderRadius: 5,
      textDecoration: "none",
      [theme.breakpoints.down("xs")]: {
        minWidth: "100%",
      },
    },
    buttonContainer: {
      [theme.breakpoints.down("xs")]: {
        position: "relative",
      },
    },
  })
);
