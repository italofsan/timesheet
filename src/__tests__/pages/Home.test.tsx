import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Home } from "../../pages/Home";
import { AuthContext } from "../../contexts/AuthContext";

afterEach(cleanup);

function renderHome(contextProps: any) {
  return render(
    <AuthContext.Provider value={contextProps}>
      <Home />
    </AuthContext.Provider>
  );
}

let contextValue = {
  user: {},
  signInWithGoogle: jest.fn(),
  signOut: jest.fn(),
  isSigned: false,
  setIsSigned: jest.fn(),
};

describe("Home page component", () => {
  it("renders correctly", () => {
    render(<Home />);

    expect(screen.getByText("Login With Google")).toBeInTheDocument();

    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should login function have been called", () => {
    renderHome(contextValue);
    const signInWithGoogleMocked = contextValue.signInWithGoogle;

    const loginButton = screen.getByRole("loginButton") as HTMLButtonElement;
    userEvent.click(loginButton);

    expect(signInWithGoogleMocked).toHaveBeenCalled();
  });
});
