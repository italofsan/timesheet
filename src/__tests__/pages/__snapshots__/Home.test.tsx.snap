// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page component renders correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-2 mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h6
          className="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
        >
          Timesheet App
        </h6>
      </div>
    </div>
  </header>
  <main
    className="makeStyles-content-4"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <div
            className="makeStyles-imageContainer-5"
          >
            <img
              alt="Time Home Page"
              className="makeStyles-image-6"
              src="time-homepage.svg"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root makeStyles-loginWrapper-7 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <div
            className="MuiPaper-root makeStyles-loginContainer-8 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <p
              className="MuiTypography-root makeStyles-loginTitle-9 MuiTypography-body1"
            >
              Log In
            </p>
            <p
              className="MuiTypography-root makeStyles-loginSubtitle-10 MuiTypography-body1"
            >
              To log in to the application you need to use a Google account
            </p>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-loginButton-11 MuiButton-containedSizeSmall MuiButton-sizeSmall"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="loginButton"
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Login With Google
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
